### ISSUE TITLE
nix for arbitrary containers?

### ISSUE THREAD

**Issue by zardus** [pwncollege/dojo#306]
We need all challenge containers to derive off of our challenge base because the tooling, and, more critically, the desktop and vscode, live there. This makes custom images a pain.

Can the nix package manager solve this problem for us? As I understand it, it's possible to have a whole system limiting under /nix. Can we just mount that in, xfce and vscode and all?

**Comment by Scoder12** [comment]
Yes, the entire system is contained in `/nix`, each package is an entry in `/nix/store` and only references other store paths. Mounting it in should work just fine. Nix can also build layered docker images.

**Comment by zardus** [comment]
A super quick test with `openvscode-server` (similar but not the same to code-server) seems to work. Didn't have enough time to figure out how to get xfce and stuff installed.

Dockerfile:

```
from ubuntu:22.04
run apt-get update && apt-get install -y nix
run nix-channel --add https://nixos.org/channels/nixos-23.11 nixos
run nix-channel --update
run nix-env -i openvscode-server
```

Copying out `/nix` and mounting it in a fresh, minimal `ubuntu` container and an `alpine` both seem to work just fine.

**Comment by Scoder12** [comment]
Should work in a `scratch` container too.
